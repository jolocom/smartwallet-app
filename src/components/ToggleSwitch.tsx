import React, { useEffect, useMemo, useRef, useState } from 'react'
import { StyleSheet, Animated } from 'react-native'
import LinearGradient from 'react-native-linear-gradient'
import { TouchableWithoutFeedback } from 'react-native-gesture-handler'

import { Colors } from '~/utils/colors'
import { isPropControlled } from '~/utils/props'

interface Props {
  onToggle?: (localSwitchState: boolean) => void
  on?: boolean
}

const ON_POSITION = 17
const OFF_POSITION = 2

const ToggleSwitch: React.FC<Props> = (props) => {
  const [isOn, setIsOn] = useState(false)

  const getOnState = () => {
    return isPropControlled(props, 'on') ? props.on : isOn
  }
  const onState = getOnState()

  const onGradientColors = [Colors.carnationPink, Colors.hyacinthPink]
  const offGradientColors = [Colors.haiti, Colors.haiti]

  const positionValue = useRef(
    new Animated.Value(onState ? ON_POSITION : OFF_POSITION),
  ).current

  const toggle = () => {
    if (isPropControlled(props, 'on')) {
      props.onToggle && props.onToggle(isOn)
    } else {
      setIsOn((prevState) => !prevState)
      props.onToggle && props.onToggle(!isOn)
    }
  }

  useEffect(() => {
    Animated.timing(positionValue, {
      toValue: onState ? ON_POSITION : OFF_POSITION,
      duration: 300,
      useNativeDriver: true,
    }).start()
  }, [onState])

  return (
    <TouchableWithoutFeedback
      testID="toggleSwitch"
      style={styles.track}
      onPressIn={toggle}
    >
      <Animated.View
        style={{
          ...styles.toggle,
          transform: [{ translateX: positionValue }],
        }}
      >
        <LinearGradient
          style={styles.gradientWrapper}
          start={{ x: 0, y: 0 }}
          end={{ x: 1, y: 0 }}
          colors={onState ? onGradientColors : offGradientColors}
        />
      </Animated.View>
    </TouchableWithoutFeedback>
  )
}

const styles = StyleSheet.create({
  track: {
    justifyContent: 'center',
    width: 45,
    height: 30,
    borderRadius: 45,
    backgroundColor: Colors.blackRussian,
  },
  toggle: {
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    width: 26,
    height: 26,
    borderRadius: 13,
    elevation: 10,
    shadowColor: Colors.black06,
    shadowOffset: {
      width: 0,
      height: 3,
    },
    shadowRadius: 1,
    shadowOpacity: 1,
  },
  gradientWrapper: {
    width: '100%',
    height: '100%',
    borderRadius: 13,
  },
})

export default ToggleSwitch
