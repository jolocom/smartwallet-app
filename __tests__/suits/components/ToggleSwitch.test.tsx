import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'

import ToggleSwitch from '~/components/ToggleSwitch'

const TOGGLE_SWITCH_ID = 'toggleSwitch'

describe('ToggleSwitch', () => {
  it('should render the switch with disabled initial state', () => {
    const { toJSON } = render(<ToggleSwitch on={false} onToggle={jest.fn()} />)

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the switch with enabled initial state', () => {
    const { toJSON } = render(<ToggleSwitch on={true} onToggle={jest.fn()} />)

    expect(toJSON()).toMatchSnapshot()
  })

  it('is controlled and should call passed onToggle fn', async () => {
    const mockFn = jest.fn()
    const { findByTestId } = render(
      <ToggleSwitch on={true} onToggle={mockFn} />,
    )

    const component = await findByTestId(TOGGLE_SWITCH_ID)
    fireEvent(component, 'pressIn')

    // expect(mockFn).toHaveBeenCalledWith(false)
    expect(mockFn).toHaveBeenCalledTimes(1)
  })
})
