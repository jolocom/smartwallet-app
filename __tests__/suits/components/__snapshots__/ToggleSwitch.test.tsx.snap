// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch should render the switch with disabled initial state 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgb(42, 34, 43)",
      "borderRadius": 45,
      "height": 30,
      "justifyContent": "center",
      "opacity": 1,
      "width": 45,
    }
  }
  testID="toggleSwitch"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 13,
        "elevation": 10,
        "height": 26,
        "justifyContent": "center",
        "position": "absolute",
        "shadowColor": "rgba(0, 0, 0, 0.06)",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
        "transform": Array [
          Object {
            "translateX": 2,
          },
        ],
        "width": 26,
      }
    }
  >
    <BVLinearGradient
      colors={
        Array [
          4278911757,
          4278911757,
        ]
      }
      endPoint={
        Object {
          "x": 1,
          "y": 0,
        }
      }
      locations={null}
      startPoint={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      style={
        Object {
          "borderRadius": 13,
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;

exports[`ToggleSwitch should render the switch with enabled initial state 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgb(42, 34, 43)",
      "borderRadius": 45,
      "height": 30,
      "justifyContent": "center",
      "opacity": 1,
      "width": 45,
    }
  }
  testID="toggleSwitch"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 13,
        "elevation": 10,
        "height": 26,
        "justifyContent": "center",
        "position": "absolute",
        "shadowColor": "rgba(0, 0, 0, 0.06)",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
        "transform": Array [
          Object {
            "translateX": 17,
          },
        ],
        "width": 26,
      }
    }
  >
    <BVLinearGradient
      colors={
        Array [
          4287699266,
          4291964265,
        ]
      }
      endPoint={
        Object {
          "x": 1,
          "y": 0,
        }
      }
      locations={null}
      startPoint={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      style={
        Object {
          "borderRadius": 13,
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;
